src/utils/daos_hdlr.c

```diff
@@ -0,0 +1,400 @@
+#include "daos_uns.h"
+
+#include "daos_hdlr.h"
+
+#define DUNS_XATTR_NAME		"user.daos"
+#define DUNS_MAX_XATTR_LEN	170
+#define DUNS_XATTR_FMT		"DAOS.%s://%36s/%36s/%s/%zu"
+
+/* TODO: implement these pool op functions
+ * int pool_list_cont_hdlr(struct cmd_args_s *ap);
+ * int pool_stat_hdlr(struct cmd_args_s *ap);
+ * int pool_get_prop_hdlr(struct cmd_args_s *ap);
+ * int pool_get_attr_hdlr(struct cmd_args_s *ap);
+ * int pool_list_attrs_hdlr(struct cmd_args_s *ap);
+ */
+
+int
+pool_query_hdlr(struct cmd_args_s *ap)
+{
+	daos_pool_info_t		 pinfo;
+	struct daos_pool_space		*ps = &pinfo.pi_space;
+	struct daos_rebuild_status	*rstat = &pinfo.pi_rebuild_st;
+	int				 i;
+	int				rc = 0;
+	int				rc2;
+
+	assert(ap != NULL);
+	assert(ap->p_op == POOL_QUERY);
+
+	rc = daos_pool_connect(ap->p_uuid, ap->group,
+			       ap->mdsrv, DAOS_PC_RO, &ap->pool,
+			       NULL /* info */, NULL /* ev */);
+	if (rc != 0) {
+		fprintf(stderr, "failed to connect to pool: %d\n", rc);
+		D_GOTO(out, rc);
+	}
+
+	pinfo.pi_bits = DPI_ALL;
+	rc = daos_pool_query(ap->pool, NULL, &pinfo, NULL, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "pool query failed: %d\n", rc);
+		D_GOTO(out_disconnect, rc);
+	}
+	D_PRINT("Pool "DF_UUIDF", ntarget=%u, disabled=%u\n",
+		DP_UUID(pinfo.pi_uuid), pinfo.pi_ntargets,
+		pinfo.pi_ndisabled);
+
+	D_PRINT("Pool space info:\n");
+	D_PRINT("- Target(VOS) count:%d\n", ps->ps_ntargets);
+	for (i = DAOS_MEDIA_SCM; i < DAOS_MEDIA_MAX; i++) {
+		D_PRINT("- %s:\n",
+			i == DAOS_MEDIA_SCM ? "SCM" : "NVMe");
+		D_PRINT("  Total size: "DF_U64"\n",
+			ps->ps_space.s_total[i]);
+		D_PRINT("  Free: "DF_U64", min:"DF_U64", max:"DF_U64", "
+			"mean:"DF_U64"\n", ps->ps_space.s_free[i],
+			ps->ps_free_min[i], ps->ps_free_max[i],
+			ps->ps_free_mean[i]);
+	}
+
+	if (rstat->rs_errno == 0) {
+		char	*sstr;
+
+		if (rstat->rs_version == 0)
+			sstr = "idle";
+		else if (rstat->rs_done)
+			sstr = "done";
+		else
+			sstr = "busy";
+
+		D_PRINT("Rebuild %s, "DF_U64" objs, "DF_U64" recs\n",
+			sstr, rstat->rs_obj_nr, rstat->rs_rec_nr);
+	} else {
+		D_PRINT("Rebuild failed, rc=%d, status=%d\n",
+			rc, rstat->rs_errno);
+	}
+
+out_disconnect:
+	/* Pool disconnect  in normal and error flows: preserve rc */
+	rc2 = daos_pool_disconnect(ap->pool, NULL);
+	if (rc2 != 0)
+		fprintf(stderr, "Pool disconnect failed : %d\n", rc2);
+
+	if (rc == 0)
+		rc = rc2;
+out:
+	return rc;
+}
+
+/* TODO implement the following container op functions
+ * all with signatures similar to this:
+ * int cont_FN_hdlr(struct cmd_args_s *ap)
+ *
+ * cont_list_objs_hdlr()
+ * int cont_query_hdlr()
+ * int cont_stat_hdlr()
+ * int cont_get_prop_hdlr()
+ * int cont_set_prop_hdlr()
+ * int cont_list_attrs_hdlr()
+ * int cont_del_attr_hdlr()
+ * int cont_get_attr_hdlr()
+ * int cont_set_attr_hdlr()
+ * int cont_create_snap_hdlr()
+ * int cont_list_snaps_hdlr()
+ * int cont_destroy_snap_hdlr()
+ * int cont_rollback_hdlr()
+ */
+
+int
+resolve_by_path(struct cmd_args_s *ap)
```

Comment:
I'm confused as why you are copying code from the duns here.
We want to move the duns "tool" and not the actual implementation / API into this generic tool.
So the duns_tool.c should be consolidated here, not the duns.c API functions. those will be needed to invoke the duns through C code rather than just the tool.

Reply:
Yes I see and have updated the patch with calls to the API, along with some reworking of the API to satisfy the use case.